Thread Network Child Device

tyler-potyondy opened the pull request:
### Pull Request Overview

This pull request adds support for a child device to successfully join a Thread network. To implement this support, I have added a ThreadNetwork capsule and made minor modifications to the UDP/6LoWPAN/15.4 capsules. 

This capsule currently only exposes one command syscall to userland. This command can be used to pass a mesh link establishment (MLE) and MAC key. After adding these keys to the Thread object, the Thread network initiates a sequence to attach the device to a parent node. This consists of the device sending a parent request -> listening for the parent response -> responding with a child id request -> listening for a child id response. With this PR, these steps all occur, and the Tock device is displayed in the child table of an OpenThread router device. This PR serves as a useful starting point to having Thread support in Tock.   

Because this PR is fairly extensive in it's scope/length, I have not tried to detail every design decision in this overview. The majority of changes and additions are in the ThreadNetwork capsule. As mentioned before, the only changes to existing capsules occur in changing the UDP capsule `driver_send_to(...)` and a minor modification in how 15.4 performs encryption.

As it stands, this pull request only implements a subset of the features and possibilities of how a child device can attach to a parent. For now, features such as the auxiliary security suite and many TLV fields are hardcoded. Overtime, these can be implemented more robustly and polished. 

### Testing Strategy

This pull request was tested by Tock successfully and consistently attaching as a child to an OpenThread router. 

### TODO or Help Wanted

This pull request still needs a number of improvements and, as mentioned before, is only a subset of the necessary functionality. More minor TODOs have been labeled in the ThreadNetwork capsule. More major TODOs include:

1) Add timeouts/retries for sending messages during child attachment to parent.
2) Support for sending UDP (fairly easy to do and fits into the existing design).
3) Control/exposure to send 15.4 packets from ThreadNetwork (likely need to add virtualizer to 15.4).
4) Send heart beats to parent (I have fully implemented all the structure/support with timers and the state machine to accomplish this. However, I need to finish writing a function to encode the payload).
5) When OpenThread responds with some 6LoWPAN packets, the entire kernel panics. This is due to some bug in the 6LoWPAN stack (not Thread), but needs to be addressed as OpenThread exposes some strange behavior.
6) Support for different security/encryption modes.

### Documentation Updated

I am leaving this unmarked for now because I suspect we need to update the docs to detail Thread networking support.

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

bradjc commented on boards/components/src/thread_network.rs:
```diff
@@ -0,0 +1,255 @@
+
+// Setup static space for the objects.
+#[macro_export]
+macro_rules! thread_network_driver_component_static {
```
```suggestion
macro_rules! thread_network_component_static {
```

Or change the struct name.

bradjc commented on capsules/extra/src/net/ipv6/ip_utils.rs:
```diff
@@ -11,6 +11,8 @@ use crate::net::ieee802154::MacAddress;
 use crate::net::ipv6::IP6Header;
 use crate::net::udp::UDPHeader;
 
+pub const DEFAULT_DST_MAC_ADDR: MacAddress = MacAddress::Short(49138);
```
Can you comment this?

tyler-potyondy replied:
Sorry for the delay getting to these comments! This default mac address is the same as the default mac address in main.rs. The underlying (and deeper issue) is described in this comment in `components/udp_mux.rs`

        // In current design, all udp senders share same IP sender, and the IP
        // sender holds the destination mac address. This means all UDP senders
        // must send to the same mac address...this works fine under the
        // assumption of all packets being routed via a single gateway router,
        // but doesn't work if multiple senders want to send to different
        // addresses on a local network. This will be fixed once we have an
        // ipv6_nd cache mapping IP addresses to dst macs
        let ip_send =
            s.4.write(capsules_extra::net::ipv6::ipv6_send::IP6SendStruct::new(
                ip6_dg,
                ipsender_virtual_alarm,
                radio_buf,
                sixlowpan_tx,
                udp_mac,
                self.dst_mac_addr,
                self.src_mac_addr,
                ip_vis,
            ));

Joining a thread child requires first sending to 0xFFFF (multicast mac address) and then sending to the specific device address of the parent. We have two solutions to move forward:

1)  My _somewhat_ hacky solution to change the function calls to add the destination mac address (currently implemented but obviously confusing with default addrs etc)
2) Change the IPV6 capsule so that the destination mac address of the radio is changed based on the IPV6 address. This would revert some of the changes made to the udp capsule and remove the need for these defaults to be used)

@ppannuto & @bradjc  please let me know your thoughts

tyler-potyondy replied:
Updating IPV6 is a much cleaner solution. I just made the changes and will push soon. This removes the need to change any UDP files and limits the changes to IPV6 to ~10 lines of code

bradjc commented on capsules/extra/src/net/ipv6/ipv6_send.rs:
```diff
@@ -107,7 +109,7 @@ pub struct IP6SendStruct<'a, A: time::Alarm<'a>> {
     sixlowpan: TxState<'a>,
     radio: &'a dyn MacDevice<'a>,
-    dst_mac_addr: MacAddress,
+    dst_mac_addr: Cell<MacAddress>,
```
Why not an OptionalCell instead of having some default?

tyler-potyondy replied:
I agree that an OptionalCell would be better here. I originally added these defaults because the kernel/capsules had some hard coded values (used for libtock-c examples). This shouldn't be an issue and I'll submit a PR updating the libtock-c examples.

bradjc commented on capsules/extra/src/net/thread/driver.rs:
A top comment in this file explaining what is intended to go above this in userspace would be quite helpful.

bradjc commented on chips/nrf52840/src/ieee802154_radio.rs:
Revert these.

ppannuto commented on boards/components/src/thread_network.rs:
````diff
@@ -0,0 +1,255 @@
+//!     .finalize(components::udp_driver_component_static!());
+//! ```
+
+// This buffer is used as an intermediate buffer for AES CCM encryption. An
+// upper bound on the required size is `3 * BLOCK_SIZE + radio::MAX_BUF_SIZE`.
````
I think this comment got detached from its buffer :)

ppannuto commented on boards/components/src/thread_network.rs:
```diff
@@ -0,0 +1,255 @@
+                crypt_buf,
+            ),
+        );
+        aes_ccm.setup(); //
```
```suggestion
        aes_ccm.setup();
```

ppannuto commented on capsules/extra/src/ieee802154/framer.rs:
```diff
@@ -751,7 +749,7 @@ impl<'a, M: Mac<'a>, A: AES128CCM<'a>> MacDevice<'a> for Framer<'a, M, A> {
         dst_addr: MacAddress,
         src_pan: PanID,
-        mut src_addr: MacAddress,
+        mut _src_addr: MacAddress,
```
Should this just be deleted?

tyler-potyondy replied:
I probably should have drawn greater attention to this. I think the short answer is yes, we should remove it. To get Thread to work, I removed the call to the lookup function (for crypto nonce) and instead just used the mac addr in the packet header. Previously, the 15.4 capsule would add a neighbor (and save the short/long addr) so that upon receiving an encrypted packet with only a short src addr, the packet could be decrypted if the neighbor had been previously saved. 

A bigger potential discussion is how should 15.4 behave? I think the answer (as we've discussed) is to strictly follow the spec. I suspect we may want to remove the neighbor functionality all together and have the 15.4 capsule follow the smallest subset of the spec.

tyler-potyondy replied:
Looking further into this today, this issue could easily be resolved by the UDP mux being given the long address when the component is created in `main.rs`. I've updated the function accordingly (now using the `src_addr` argument)

ppannuto commented on capsules/extra/src/net/thread/thread_utils.rs:
```diff
@@ -0,0 +1,282 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
```
```suggestion
// Copyright Tock Contributors 2023.
```

Here and elsewhere :)

ppannuto commented on capsules/extra/src/net/thread/thread_utils.rs:
```diff
@@ -0,0 +1,282 @@
+    IPAddr(output)
+}
+
+pub fn mac_from_ipv6(ipv6: IPAddr) -> [u8; 8] {
+    // Helper function to generate the mac address from the mac address;
+    // reversing the tranformation used/described in `generate_src_ipv6`
```
```suggestion
/// Helper function to generate the mac address from the mac address;
/// the tranformation used/described in `generate_src_ipv6`
pub fn mac_from_ipv6(ipv6: IPAddr) -> [u8; 8] {
```

Here and elsewhere, promote these to doc comments [especially important for `pub` functions]

ppannuto reviewed (commented):
Quick / partial cosmetic pass while definitely not multitasking 😇

bradjc commented on capsules/extra/src/net/thread/driver.rs:
Would a userspace app using this driver work if the same app was loaded twice? If not, this driver should give priority to the first app and return errors to all others.

tyler-potyondy replied:
Long term this should be handled more robustly. For now I think it makes the most sense to "lock" Thread to whatever userspace app first calls it. I just pushed this change.

brghena commented on capsules/extra/src/ieee802154/driver.rs:
```diff
@@ -505,15 +518,22 @@ impl framer::DeviceProcedure for RadioDriver<'_> {
+        });
+        if res.is_none() {
+            self.backup_device_procedure
+                .unwrap_or_panic()
```
Can you give some more details here, maybe in a comment. Why is this panic okay? What does it mean to do a "backup procedure"? Is this something that is quite possible to fail at runtime, or is this supposed to always succeed, unless there's an implementation issue?

tyler-potyondy replied:
The device backup procedure arose due to the current design of 15.4. The 15.4 driver stores the network keys and the framer looks them up accordingly using a lookup procedure trait (when receiving/sending a 15.4 packet). Once joined to the Thread network, Thread messages are encrypted within the UDP payload and also at the link layer. Because of this, we need to possess a way to add to the Radio Driver's stored keys. My solution was to add a "backup lookup" so that if a key was not found, the 15.4 framer would check for the Thread key next. Long term it is not ideal for the 15.4 driver to hold these keys (since it makes it rather clunky for other capsules to add/remove keys).

This is something I will keep thinking about. In the short term, my fix should be adequet and does not cause much disruption to the 15.4 capsule. I will go ahead and remove the `panic` because it should likely just return an error if this is not set during the configuration. 

I will add some comments and TODOs for this so we have a note moving forward.

brghena commented on capsules/extra/src/ieee802154/framer.rs:
```diff
@@ -450,11 +441,18 @@ impl<'a, M: Mac<'a>, A: AES128CCM<'a>> Framer<'a, M, A> {
+                                MacAddress::Long(val) => val,
+                                MacAddress::Short(_) => {
+                                    kernel::debug!("[15.4] DROPPED PACKET - error only short address provided on encrypted packet.");
+                                    return None
+                                },
+                            },
                             None => {
-                                return None;
-                            }
+                                kernel::debug!("[15.4] DROPPED PACKET - Malformed, no src address provided.");
+                                return None
+                            },
```
@amit or @hudson-ayers or maybe @bradjc: Do we have a policy in Tock about "permanent" debug statements in capsules?

I think I'm generally against them, but if other capsules are doing it anyways, it doesn't hurt here.

brghena commented on capsules/extra/src/ieee802154/mac.rs:
```diff
@@ -172,20 +171,21 @@ impl<'a, R: radio::Radio<'a>> radio::RxClient for AwakeMac<'a, R> {
-
         if addr_match {
-            //debug!("[AwakeMAC] Rcvd a 15.4 frame addressed to this device");
+            // debug!("[AwakeMAC] Rcvd a 15.4 frame addressed to this device");
```
Maybe this is a good compromise: leave the debug statements there, but commented out. They're essentially comments, but you can activate them again if you are debugging.

ppannuto replied:
I think I actually like the code as-is — the common-case / correct-case debug statements are commented out, so in normal operation the console isn't flooded. Elsewhere, when errors / uncommon cases occur, debugs are still printed by default.

brghena commented on capsules/extra/src/net/thread/driver.rs:
```diff
@@ -0,0 +1,744 @@
+                .bind(socket, THREAD_PORT_NUMBER, self.net_cap)
+            {
+                Ok((udp_tx, udp_rx)) => (udp_tx, udp_rx),
+                Err(sock) => panic!("failed binding of port to socket - {:?}", sock),
+            },
+            Err(return_code) => panic!("Error in retrieving socket - {:?}", return_code),
```
Again, do we think these are quite plausible runtime issues, or only triggered if the implementation itself is bad?

Generally, my thought is that we shouldn't be calling `panic` from a capsule for runtime issues. But a `panic` that would only occur if configuration is wrong or if there's an implementation bug, seems reasonable.

tyler-potyondy replied:
I agree with your thinking about only calling `panic` on configuration issues. This case you cited here interestingly spans both runtime / configuration. Thread will be unable to operate if it cannot bind to port 19788 for UDP (unable to send MLE packets).   `init_thread_binding()` is called in the Thread component and binds Thread to the needed UDP port. This is predominantly a configuration case (since it is occurring in components). Rethinking this now,  we can move this functionality to entirely exist in the component. 

This brings up an interesting design decision we should discuss in the next networking wg. 

> Should Thread be the only entity capable of binding to 19788?
_(I lean away from this because this seems needlessly restrictive and we can bind Thread to this port once an application requests to create or join a Thread network)_

In the short term and for this pull request, I think this binding should be moved to the component and the error checking removed. The UDP send request will fail if this binding does not occur, and this case is handled in the Thread implementation (terminates the Thread joining).

tyler-potyondy replied:
Correction: UDP `driver_send_to(...)` does not require the port to be specified to send udp packets. This means Thread will not be able to receive packets and silently fail/eventually time out if this binding somehow fails. Once again, I think this is fine for now, but long term will need a more robust error handling.

brghena commented on capsules/extra/src/net/thread/driver.rs:
```diff
@@ -0,0 +1,744 @@
+                self.sender
+                    .driver_send_to(
+                        dest_ipv6,
+                        19788,
```
This feels like it should be a named thing, rather than just a raw number.

brghena replied:
Oh, it looks like you even have one: THREAD_PORT_NUMBER

brghena reviewed (changes requested):
Some thoughts for you

ppannuto commented on capsules/extra/src/ieee802154/driver.rs:
```diff
@@ -522,12 +542,20 @@ impl framer::KeyProcedure for RadioDriver<'_> {
-        })
+        });
+
+        if res.is_none() {
+            self.backup_key_procedure
+                .unwrap_or_panic()
+                .lookup_key(SecurityLevel::EncMic32, KeyId::Index(2))
```
I _think_ it'd be "rusty-er" to put the body of this `if` as closure of a chained `.unwrap_or_else(|` and omit the `res` variable.

ppannuto commented on capsules/extra/src/net/ipv6/ipv6_send.rs:
```diff
@@ -141,12 +142,28 @@ impl<'a, A: time::Alarm<'a>> IP6Sender<'a> for IP6SendStruct<'a, A> {
+            dst_mac_addr = self.dst_mac_addr;
+        }
+
+        let _ = self
```
I see that this is just a refactor of what was there before, but maybe a TODO for a future PR to do something with the error / return here. Could be worth dropping a `// TODO`, or could just ignore me here :)

ppannuto reviewed (commented):
I think this is just about ready to go, only a couple of small things I saw

alevy commented:
@tyler-potyondy can you maybe summarize where this PR stands?

It's been open for over a month, and has had significant discussion/back-and-forth. My sense is that this is ready to go, but would be useful to have a summary comment for review.

In addition, my sense is that the changes in this PR are almost entirely contained to a Thread-specific capsule, with some bug fixes in 6lowpan, and bleeding out into one board and a component (so really minor changes there). In other words, as long as the stakeholders involved in Thread are happy with this, and the 6lowpan fix seems right, we should merge this in service of progress on Thread, even if there are some design questions outstanding (e.g., regarding whether/how to enforce port binding more broadly).

tyler-potyondy commented:
@alevy PR Status Summary:

The PR (from my perspective) is complete and ready to merge. All comments/reviews from @bradjc @ppannuto @brghena have been addressed and implemented. 

The PR is almost entirely contained to the Thread capsule (and a Thread component) making it low impact to the rest of Tock. As it stands, the Thread capsule provides support to join a Thread network as a child device. Minor changes have been made to the IPv6 and 802.15.4 capsules.  

Previously, IPv6 would exclusively send packets to the destination MAC address contained in the IPv6 sending struct. Because Thread requires the ability to send to multiple MAC addresses, a minor modification was required to IPv6 that would only alter IPv6's behavior when sending Thread messages (I added a comment addressing this assumption as this assumption could potentially cause issues in a future use case of the networking stack). 

The changes to the 15.4 capsule relate primarily to multicast addresses being passed to higher layers (otherwise packets are dropped as they do not match the device MAC) and altering the current network key storage. The alterations to the network key storage allow Thread a mechanism to add the Thread MAC key for decrypting link layer encrypted packets. This change will require me to update the `libtock-c` 15.4 examples (_TODO_).
